clang/lib/Sema/SemaExprCXX.cpp

```diff
@@ -7939,7 +7927,14 @@ ExprResult Sema::ActOnStartCXXMemberReference(Scope *S, Expr *Base,
           }
         }
         return ExprError();
+      } else if (Result.isUnset()) {
+        // BuildOverloadedArrowExpr returns an empty expression to indicate
+        // that we need to build a dependent overloaded arrow expression.
```

Comment:
I have 2 concerns with that
 1/ It's a bit too clever of an interface
 2/ We should make sure this can only happens in Sema, right? I wonder if we need an ActOnOverloadedArrowExpr to wrap that logic such that (it would duplicate the lookup code but i think that's a better outcome)

WDYT?

Reply:
> I have 2 concerns with that 
> 1/ It's a bit too clever of an interface

Although I agree it's an imperfect interface, this is more of a "transitional" fix. Ideally we will build the full AST for calls to members of the current instantiation in the future. For now, the usage of `ExprEmpty` in this patch isn't unfounded so I think it's "good enough". Alternatively I can add a `bool& IsDependent` out parameter to implement this functionality. 

> 2/ We should make sure this can only happens in Sema, right? I wonder if we need an ActOnOverloadedArrowExpr to wrap that logic such that (it would duplicate the lookup code but i think that's a better outcome)

I'm not entirely sure what this means. Are you saying that we should expect the result to _not_ be `ExprEmpty` during instantiation?

Reply:
I'm not sure RebuildCXXOperatorCallExpr and its caller would deal with an ExprEmpty when calling  
ActOnOverloadedArrowExpr

So yeah, maybe output parameter + assert in treetransform would be better?

Reply:
@cor3ntin Pushed changes that do what you described